solr/core/src/java/org/apache/solr/update/processor/ConditionalUpsertProcessorFactory.java

```diff
@@ -0,0 +1,134 @@
+    // Ensure the parameters are forwarded to the leader
+    DistributedUpdateProcessorFactory.addParamToDistributedRequestWhitelist(req, PARAM_IGNORE_CONDITIONAL_UPSERTS);
+
+    // there may be cases where we want to revert to "regular" behaviour without the conditional
+    // upsert logic interfering
+    boolean ignoreConditionalUpserts = req.getOriginalParams().getBool(PARAM_IGNORE_CONDITIONAL_UPSERTS, false);
+
+    return new ConditionalUpsertUpdateProcessor(req, next, conditions, ignoreConditionalUpserts);
+  }
+
+  @Override
+  public void inform(SolrCore core) {
+    if (core.getUpdateHandler().getUpdateLog() == null) {
+      throw new SolrException(SERVER_ERROR, "updateLog must be enabled.");
+    }
+
+    if (core.getLatestSchema().getUniqueKeyField() == null) {
+      throw new SolrException(SERVER_ERROR, "schema must have uniqueKey defined.");
+    }
+  }
+
+  static class ConditionalUpsertUpdateProcessor extends UpdateRequestProcessor {
+    private final SolrCore core;
+
+    private DistributedUpdateProcessor distribProc;  // the distributed update processor following us
+    private DistributedUpdateProcessor.DistribPhase phase;
+    private final List<UpsertCondition> conditions;
+    private final boolean ignoreConditionalUpserts;
+
+    ConditionalUpsertUpdateProcessor(SolrQueryRequest req,
+                                     UpdateRequestProcessor next,
+                                     List<UpsertCondition> conditions,
+                                     boolean ignoreConditionalUpserts) {
+      super(next);
+      this.core = req.getCore();
+      this.conditions = conditions;
+      this.ignoreConditionalUpserts = ignoreConditionalUpserts;
+
+      for (UpdateRequestProcessor proc = next ;proc != null; proc = proc.next) {
+        if (proc instanceof DistributedUpdateProcessor) {
+          distribProc = (DistributedUpdateProcessor)proc;
+          break;
+        }
+      }
+
+      if (distribProc == null) {
+        throw new SolrException(SolrException.ErrorCode.SERVER_ERROR, "DistributedUpdateProcessor must follow ConditionalUpsertProcessor");
+      }
+
+      phase = DistributedUpdateProcessor.DistribPhase.parseParam(req.getParams().get(DISTRIB_UPDATE_PARAM));
+    }
+
+    boolean isLeader(UpdateCommand cmd) {
+      if ((cmd.getFlags() & (UpdateCommand.REPLAY | UpdateCommand.PEER_SYNC)) != 0) {
+        return false;
+      }
+      if (phase == DistributedUpdateProcessor.DistribPhase.FROMLEADER) {
+        return false;
+      }
+      return distribProc.isLeader();
+    }
+
+    @Override
+    public void processAdd(AddUpdateCommand cmd) throws IOException {
+      if (!ignoreConditionalUpserts && !conditions.isEmpty() && isLeader(cmd)) {
+        BytesRef indexedDocId = cmd.getIndexedId();
+        SolrInputDocument oldDoc = RealTimeGetComponent.getInputDocument(core, indexedDocId, RealTimeGetComponent.Resolution.DOC);
```

Comment:
I wonder if [this commit Tim did](https://github.com/BrandwatchLtd/bw-lucene-solr/commit/eccb3e52961698db1e3b4778810a87003d0349ee) would need to be applied here? Without opening this in intellij / looking at the java docs for the `getInputDocument` I don't know for sure, but Tim did have to add some extra parameters for the 8.11.2 branch with the commit being:
> Fixes for solr8 build

Reply:
yeah this was essentially the difference I was talking about.  I'd just chose a difference method for my fix.